src/python/pants/util/frozendict_test.py

```diff
@@ -141,3 +142,27 @@ class Mutable:
         FrozenDict({Mutable(0): "a"})
     with pytest.raises(TypeError):
         FrozenDict({"a": Mutable(0)})
+
+
+def test_lazy_frozen_dict() -> None:
+    loaded: DefaultDict[str, int] = defaultdict(int)
+
+    def load_value(s: str) -> str:
+        loaded[s] += 1
+        return "".join(reversed(s))
+
+    ld1 = LazyFrozenDict({"a": lambda: load_value("1234"), "b": lambda: load_value("abcd")})
+    hashvalue = hash(ld1)
+    assert len(ld1) == 2
+    assert len(loaded) == 0
+
+    # Test memoization, that we don't invoke the loader twice.
```

Comment:
To double check...have you tried removing the memoized decorator and confirming the test fails without it? I'm pretty sure this test is written correctly, and I like how you did it. But to be extra confident.

Reply:
True, I should've done that, but hadn't. Tried it now, and you're right, it works:
```
>       assert loaded == {"abcd": 1}
E       AssertionError: assert equals failed
E         defaultdict(<class 'int'>, {  {'abcd': 1}                  
E           'abcd': 2,                                               
E         })
```